### DISCUSSION TITLE
Question about cycles

### DISCUSSION THREAD

**Discussion by trantor** [cue-lang/cue#1520] [Q&A]
Hello.
I've been trying to achieve what I'll describe below with cue.

Let's take a YAML structure such as the following:
```
vhosts:
  - name: vhost1
    auth: true
  - name: vhost2
    auth: false
auth_enabled:
  attr: whatever
```

Now the idea would be to make `auth_enabled` optional if none of `vhosts` has `auth` set to `true` and mandatory if at least one has.

My attempt was through this cue code, taking inspiration from the tutorials:
```
import "list"

#host_def

#host_def: this={
  vhosts: [...#vhost]
  #auth_present: list.MinItems([for k,v in this.vhosts if v.auth { v.auth }],1)
  if #auth_present {
    auth_enabled: "attr" != ""
  }
  if !#auth_present {
    auth_enabled?: "attr" != ""
  }

}

#vhost: {
  name: string
  auth: bool
}
```
If I try to `vet` I get `cycle error`, and `eval` gives no output whatsoever.

I worked around the problem, somewhat, and got somewhere with this
```
import "list"

#host_def

#host_def: this={
  vhosts: [...#vhost]
  auth_enabled: *[for k,v in this.vhosts if v.auth { { "attr" != "" } }] | {...}
}

#vhost: {
  name: string
  auth: bool
}
```
but that's less precise and it does not allow me to make `auth_enabled` optional at all, obviously.

Is this a result of a lack of understanding on my account or is it a bug? (That would make two I stumble upon on the first day of trying cue, so someone should sponsor me as a bug hunter in that case :rofl: )

**Comment by eonpatapon** [comment]
The cycle error seems to happen because of the `this` alias.

But in your case you don't need it. This is valid:

```cue
import "list"

#host_def

#host_def: {
	vhosts: [...#vhost]
	#auth_present: list.MinItems([ for k, v in vhosts if v.auth {v.auth}], 1)
	if #auth_present {
		auth_enabled: "attr" != ""
	}
	if !#auth_present {
		auth_enabled?: "attr" != ""
	}
}

#vhost: {
	name: string
	auth: bool
}
```

Not sure what you intended to do with `auth_enabled: "attr" != ""` though.
